File: src/my_project/medium_problems/from101to150/maximum_width_tree_BFS.py
from typing import List, Optional 
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_width = 0
        q = deque([(root, 0)])
        while q:
            length = len(q)
            max_width = max(max_width, q[-1][1] - q[0][1] + 1)
            for _ in range(length):
                node, x = q.popleft()
                if node.left:
                    q.append((node.left, 2 * x))
                if node.right:
                    q.append((node.right, 2 * x + 1))
                    
        return max_width

File: my_project_typescript/medium_problems_typescript/from51to100/maximumWidthTreeBFS.ts

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}


function widthOfBinaryTree(root: TreeNode | null): number {

    let maxWidth = 0;
    if (!root) return maxWidth;
    const queue: [TreeNode, number][] = [[root, 0]]; // TreeNode, Position
    while (queue.length) {
      const levelWidth = queue[queue.length - 1][1] - queue[0][1] + 1;
      maxWidth = Math.max(maxWidth, levelWidth);
      const len = queue.length;
      for (let i = 0; i < len; i++) {
        const [node, pos] = queue.shift()!;
        if (node.left) {
          queue.push([node.left, pos * 2]);
        }
        if (node.right) {
          queue.push([node.right, pos * 2 + 1]);
        }
      }
    }
    return maxWidth;

};